File: packages/contracts/python/noa_contracts/geometry_rule_set.py
from __future__ import annotations

from typing import Literal, Union
from pydantic import BaseModel


class GlazingRatioRule(BaseModel):
    rule_type: Literal["glazing_ratio"] = "glazing_ratio"
    target_ratio: float
    applies_to: Literal["exterior_walls", "all_walls"]
    confidence: float


class RoofRule(BaseModel):
    rule_type: Literal["roof_form"] = "roof_form"
    form: Literal["flat", "shed", "gabled", "hipped", "butterfly"]
    pitch_degrees: float | None
    confidence: float


class OverhangRule(BaseModel):
    rule_type: Literal["overhang"] = "overhang"
    model_config = {"extra": "allow"}


class FacadeGridRule(BaseModel):
    rule_type: Literal["facade_grid"] = "facade_grid"
    model_config = {"extra": "allow"}


class FloorHeightRule(BaseModel):
    rule_type: Literal["floor_height"] = "floor_height"
    model_config = {"extra": "allow"}


class MaterialHintRule(BaseModel):
    rule_type: Literal["material_hint"] = "material_hint"
    model_config = {"extra": "allow"}


GeometryRule = Union[
    GlazingRatioRule, OverhangRule, RoofRule,
    FacadeGridRule, FloorHeightRule, MaterialHintRule,
]


class GeometryRuleSet(BaseModel):
    schema_version: Literal["1.0"] = "1.0"
    id: str
    style_profile_id: str
    derived_at: str
    confidence_overall: float
    rules: list[GeometryRule]

File: packages/contracts/v1/geometry-rule-set.ts
import type { UUID, ISO8601 } from "./common.js";

export interface GlazingRatioRule {
  rule_type: "glazing_ratio";
  target_ratio: number;
  applies_to: "exterior_walls" | "all_walls";
  confidence: number;
}

export interface OverhangRule {
  rule_type: "overhang";
  [key: string]: unknown;
}

export interface RoofRule {
  rule_type: "roof_form";
  form: "flat" | "shed" | "gabled" | "hipped" | "butterfly";
  pitch_degrees: number | null;
  confidence: number;
}

export interface FacadeGridRule {
  rule_type: "facade_grid";
  [key: string]: unknown;
}

export interface FloorHeightRule {
  rule_type: "floor_height";
  [key: string]: unknown;
}

export interface MaterialHintRule {
  rule_type: "material_hint";
  [key: string]: unknown;
}

export type GeometryRule =
  | GlazingRatioRule
  | OverhangRule
  | RoofRule
  | FacadeGridRule
  | FloorHeightRule
  | MaterialHintRule;

export interface GeometryRuleSet {
  schema_version: "1.0";
  id: UUID;
  style_profile_id: UUID;
  derived_at: ISO8601;
  confidence_overall: number;
  rules: GeometryRule[];
}
